File: Enclave For Windows/src/EnclaveCore/EnclaveFolder.cs
using System.Buffers.Binary;
using System.Text;

namespace EnclaveCore;

/// <summary>
/// Folder packing/unpacking. A directory tree is serialized into an ENCLPKG1
/// bundle (paths stored '/'-separated, platform-neutral) which is then encrypted
/// as a single archive. Byte-compatible with the macOS build, so a folder packed
/// on one platform unpacks on the other.
/// </summary>
public static class EnclaveFolder
{
    public static readonly byte[] Magic = Encoding.ASCII.GetBytes("ENCLPKG1");
    public const byte Version = 1;
    public const string FilenameSuffix = ".enclavefolder";
    public const int MaxEntries = 100_000;
    public const int MaxPathLength = 4_096;
    public static long MaxEntrySize => EnclaveCrypto.MaxPlaintextSize;

    private static readonly UTF8Encoding StrictUtf8 = new(false, throwOnInvalidBytes: true);

    public static bool IsDirectory(string path) => Directory.Exists(path);

    public static bool IsFolderPayload(ReadOnlySpan<byte> data) => data.StartsWith(Magic);

    public static string StoredFilename(string directoryPath) =>
        new DirectoryInfo(directoryPath).Name + FilenameSuffix;

    public static bool IsFolderStoredName(string filename)
    {
        string baseName = LastComponent(filename);
        return baseName.EndsWith(FilenameSuffix, StringComparison.Ordinal);
    }

    public static string DisplayFolderName(string storedFilename)
    {
        string baseName = LastComponent(storedFilename);
        if (baseName.EndsWith(FilenameSuffix, StringComparison.Ordinal))
            return baseName[..^FilenameSuffix.Length];
        return baseName;
    }

    /// <summary>Visible archive name when filename encryption is off (no internal suffix).</summary>
    public static string PlaintextArchiveName(string storedFilename)
    {
        if (IsFolderStoredName(storedFilename))
            return DisplayFolderName(storedFilename);
        return LastComponent(storedFilename);
    }

    // ===================== Pack =====================

    public static byte[] Pack(string directoryPath)
    {
        string root = Path.GetFullPath(directoryPath);
        if (!IsDirectory(root)) throw EnclaveException.InvalidFormat();

        var entries = new List<(string Path, byte[] Data)>();
        long payloadSize = Magic.Length + 1 + 4;

        foreach (string file in EnumerateRegularFiles(root))
        {
            long size;
            try { size = new FileInfo(file).Length; }
            catch { throw EnclaveException.UnreadableFile(Path.GetFileName(file)); }
            if (size > MaxEntrySize) throw EnclaveException.PayloadTooLarge();

            string relativePath = RelativePath(root, file);
            byte[] fileData = File.ReadAllBytes(file);
            if (fileData.LongLength > MaxEntrySize) throw EnclaveException.PayloadTooLarge();

            long entrySize = 4 + Encoding.UTF8.GetByteCount(relativePath) + 8 + fileData.Length;
            if (payloadSize + entrySize > EnclaveCrypto.MaxPlaintextSize)
                throw EnclaveException.PayloadTooLarge();
            payloadSize += entrySize;
            entries.Add((relativePath, fileData));

            if (entries.Count > MaxEntries) throw EnclaveException.PayloadTooLarge();
        }

        if (entries.Count == 0) throw EnclaveException.EmptyFolder();

        entries.Sort((a, b) => string.CompareOrdinal(a.Path, b.Path));

        var archive = new MemoryStream();
        archive.Write(Magic);
        archive.WriteByte(Version);
        WriteUInt32(archive, (uint)entries.Count);
        foreach (var entry in entries)
            AppendEntry(archive, entry.Path, entry.Data);

        return archive.ToArray();
    }

    // ===================== Unpack =====================

    public static void Unpack(byte[] payload, string parentDirectory, string folderName)
    {
        int offset = 0;
        if (!IsFolderPayload(payload)) throw EnclaveException.InvalidFormat();
        offset = Magic.Length;

        if (offset + 1 + 4 > payload.Length) throw EnclaveException.InvalidFormat();
        byte fileVersion = payload[offset];
        offset += 1;
        if (fileVersion != Version) throw EnclaveException.InvalidFormat();

        long entryCount = ReadUInt32(payload, ref offset);
        if (entryCount < 1 || entryCount > MaxEntries) throw EnclaveException.InvalidFormat();

        string safeFolderName = SanitizeFolderName(folderName);
        string destination = Path.GetFullPath(Path.Combine(Path.GetFullPath(parentDirectory), safeFolderName));
        EnsureDestinationAvailable(destination);

        for (long i = 0; i < entryCount; i++)
        {
            var (relativePath, fileData) = ReadEntry(payload, ref offset);
            string safePath = SanitizeRelativePath(relativePath);

            string[] parts = safePath.Split('/');
            var combineArgs = new string[parts.Length + 1];
            combineArgs[0] = destination;
            Array.Copy(parts, 0, combineArgs, 1, parts.Length);
            string fileFull = Path.GetFullPath(Path.Combine(combineArgs));

            if (!IsContainedInDirectory(fileFull, destination))
                throw EnclaveException.InvalidFormat();

            string? parent = Path.GetDirectoryName(fileFull);
            if (!string.IsNullOrEmpty(parent))
                Directory.CreateDirectory(parent);
            File.WriteAllBytes(fileFull, fileData);
        }

        if (offset != payload.Length) throw EnclaveException.InvalidFormat();
    }

    private static void EnsureDestinationAvailable(string destination)
    {
        if (Directory.Exists(destination))
        {
            if (Directory.EnumerateFileSystemEntries(destination).Any())
                throw EnclaveException.DestinationExists();
            return;
        }
        if (File.Exists(destination))
            throw EnclaveException.DestinationExists();
        Directory.CreateDirectory(destination);
    }

    // ===================== Path safety =====================

    private static string SanitizeFolderName(string name)
    {
        if (string.IsNullOrEmpty(name) || name.Contains('/') || name.Contains('\\'))
            throw EnclaveException.InvalidFormat();
        string baseName = LastComponent(name);
        if (string.IsNullOrEmpty(baseName) || baseName == "." || baseName == "..")
            throw EnclaveException.InvalidFormat();
        if (baseName.Contains(".."))
            throw EnclaveException.InvalidFormat();
        return baseName;
    }

    private static string RelativePath(string root, string file)
    {
        string rel = Path.GetRelativePath(root, file).Replace('\\', '/');
        return SanitizeRelativePath(rel);
    }

    private static string SanitizeRelativePath(string path)
    {
        if (string.IsNullOrEmpty(path) || path.StartsWith('/') || path.Contains('\\'))
            throw EnclaveException.InvalidFormat();
        if (Encoding.UTF8.GetByteCount(path) > MaxPathLength)
            throw EnclaveException.InvalidFormat();

        string[] parts = path.Split('/');
        foreach (string part in parts)
        {
            if (part.Length == 0 || part == "." || part == "..")
                throw EnclaveException.InvalidFormat();
        }
        return string.Join('/', parts);
    }

    private static bool IsContainedInDirectory(string fileFullPath, string directoryFullPath)
    {
        string normalizedDir = directoryFullPath.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
        string rel = Path.GetRelativePath(normalizedDir, fileFullPath);
        if (rel == "." || rel == "..") return false;
        if (rel.StartsWith(".." + Path.DirectorySeparatorChar, StringComparison.Ordinal)) return false;
        if (rel.StartsWith("../", StringComparison.Ordinal)) return false;
        if (Path.IsPathRooted(rel)) return false;
        return true;
    }

    private static string LastComponent(string name)
    {
        int slash = name.LastIndexOfAny(new[] { '/', '\\' });
        return slash >= 0 ? name[(slash + 1)..] : name;
    }

    /// <summary>Recursive walk that skips symlinks/junctions (reparse points),
    /// hidden files, and dotfiles — mirroring the macOS enumerator options and
    /// avoiding traversal out of the tree via a junction.</summary>
    private static IEnumerable<string> EnumerateRegularFiles(string root)
    {
        var stack = new Stack<string>();
        stack.Push(root);
        while (stack.Count > 0)
        {
            string dir = stack.Pop();
            string[] children;
            try { children = Directory.GetFileSystemEntries(dir); }
            catch { continue; }

            foreach (string entry in children)
            {
                FileAttributes attr;
                try { attr = File.GetAttributes(entry); }
                catch { continue; }

                string name = Path.GetFileName(entry);
                if (name.StartsWith('.')) continue;
                if ((attr & FileAttributes.Hidden) != 0) continue;
                if ((attr & FileAttributes.ReparsePoint) != 0) continue;

                if ((attr & FileAttributes.Directory) != 0)
                {
                    stack.Push(entry);
                    continue;
                }
                yield return entry;
            }
        }
    }

    // ===================== Wire helpers =====================

    private static void WriteUInt32(Stream stream, uint value)
    {
        Span<byte> buf = stackalloc byte[4];
        BinaryPrimitives.WriteUInt32BigEndian(buf, value);
        stream.Write(buf);
    }

    private static uint ReadUInt32(byte[] data, ref int offset)
    {
        if (offset + 4 > data.Length) throw EnclaveException.InvalidFormat();
        uint value = BinaryPrimitives.ReadUInt32BigEndian(data.AsSpan(offset, 4));
        offset += 4;
        return value;
    }

    private static void AppendEntry(Stream stream, string path, byte[] fileData)
    {
        byte[] pathData = Encoding.UTF8.GetBytes(path);
        WriteUInt32(stream, (uint)pathData.Length);
        stream.Write(pathData);
        Span<byte> size = stackalloc byte[8];
        BinaryPrimitives.WriteUInt64BigEndian(size, (ulong)fileData.LongLength);
        stream.Write(size);
        stream.Write(fileData);
    }

    private static (string, byte[]) ReadEntry(byte[] data, ref int offset)
    {
        if (data.Length - offset < 4) throw EnclaveException.InvalidFormat();

        long pathLength = ReadUInt32(data, ref offset);
        if (pathLength == 0 || pathLength > MaxPathLength || offset + pathLength + 8 > data.Length)
            throw EnclaveException.InvalidFormat();

        byte[] pathData = data[offset..(offset + (int)pathLength)];
        offset += (int)pathLength;
        string? path = TryUtf8(pathData);
        if (path is null) throw EnclaveException.InvalidFormat();

        ulong fileLength = BinaryPrimitives.ReadUInt64BigEndian(data.AsSpan(offset, 8));
        offset += 8;
        // Zero-length entries are valid — empty files round-trip faithfully.
        if (fileLength > (ulong)MaxEntrySize) throw EnclaveException.InvalidFormat();
        int len = (int)fileLength;
        if (offset + len > data.Length) throw EnclaveException.InvalidFormat();

        byte[] fileData = data[offset..(offset + len)];
        offset += len;
        return (path, fileData);
    }

    private static string? TryUtf8(byte[] bytes)
    {
        try { return StrictUtf8.GetString(bytes); }
        catch { return null; }
    }
}

File: Enclave-macOS/Sources/EnclaveFolder.swift
import Foundation

enum EnclaveFolder {
    static let magic = Data("ENCLPKG1".utf8)
    static let version: UInt8 = 1
    static let filenameSuffix = ".enclavefolder"
    static let maxEntries = 100_000
    static let maxPathLength = 4_096
    static let maxEntrySize = EnclaveCrypto.maxPlaintextSize

    static func isDirectory(_ url: URL) -> Bool {
        var isDirectory: ObjCBool = false
        return FileManager.default.fileExists(atPath: url.path, isDirectory: &isDirectory) && isDirectory.boolValue
    }

    static func isFolderPayload(_ data: Data) -> Bool {
        data.starts(with: magic)
    }

    static func storedFilename(forDirectory url: URL) -> String {
        url.lastPathComponent + filenameSuffix
    }

    static func isFolderStoredName(_ filename: String) -> Bool {
        let base = (filename as NSString).lastPathComponent
        return base.hasSuffix(filenameSuffix)
    }

    static func displayFolderName(from storedFilename: String) -> String {
        let base = (storedFilename as NSString).lastPathComponent
        if base.hasSuffix(filenameSuffix) {
            return String(base.dropLast(filenameSuffix.count))
        }
        return base
    }

    /// Visible archive filename when filename encryption is off (no internal suffix).
    static func plaintextArchiveName(from storedFilename: String) -> String {
        if isFolderStoredName(storedFilename) {
            return displayFolderName(from: storedFilename)
        }
        return (storedFilename as NSString).lastPathComponent
    }

    static func pack(directory url: URL) throws -> Data {
        let root = url.standardizedFileURL
        guard isDirectory(root) else {
            throw EnclaveError.invalidFormat
        }

        guard let enumerator = FileManager.default.enumerator(
            at: root,
            includingPropertiesForKeys: [.isRegularFileKey, .isSymbolicLinkKey, .fileSizeKey],
            options: [.skipsPackageDescendants, .skipsHiddenFiles]
        ) else {
            throw EnclaveError.unreadableFile(root.lastPathComponent)
        }

        var entries: [(path: String, data: Data)] = []
        var payloadSize = magic.count + 1 + 4

        while let item = enumerator.nextObject() as? URL {
            let values = try item.resourceValues(forKeys: [.isRegularFileKey, .isSymbolicLinkKey, .fileSizeKey])
            if values.isSymbolicLink == true {
                continue
            }
            guard values.isRegularFile == true else {
                continue
            }

            if let fileSize = values.fileSize, fileSize > maxEntrySize {
                throw EnclaveError.payloadTooLarge
            }

            let relativePath = try relativePath(from: root, to: item)
            let fileData = try Data(contentsOf: item)
            guard fileData.count <= maxEntrySize else {
                throw EnclaveError.payloadTooLarge
            }

            let entrySize = 4 + relativePath.utf8.count + 8 + fileData.count
            guard payloadSize + entrySize <= EnclaveCrypto.maxPlaintextSize else {
                throw EnclaveError.payloadTooLarge
            }
            payloadSize += entrySize
            entries.append((relativePath, fileData))

            guard entries.count <= maxEntries else {
                throw EnclaveError.payloadTooLarge
            }
        }

        guard !entries.isEmpty else {
            throw EnclaveError.emptyFolder
        }

        entries.sort { $0.path < $1.path }

        var archive = Data()
        archive.append(magic)
        archive.append(version)
        appendUInt32(UInt32(entries.count), to: &archive)

        for entry in entries {
            appendBlob(entry.path, data: entry.data, to: &archive)
        }

        return archive
    }

    static func unpack(payload: Data, to parentDirectory: URL, folderName: String) throws {
        var offset = 0
        guard payload.starts(with: magic) else {
            throw EnclaveError.invalidFormat
        }
        offset = magic.count

        guard offset + 1 + 4 <= payload.count else {
            throw EnclaveError.invalidFormat
        }
        let fileVersion = payload[offset]
        offset += 1
        guard fileVersion == version else {
            throw EnclaveError.invalidFormat
        }

        let entryCount = Int(try readUInt32(payload, offset: &offset))
        guard entryCount >= 1, entryCount <= maxEntries else {
            throw EnclaveError.invalidFormat
        }

        let safeFolderName = try sanitizeFolderName(folderName)
        let destination = parentDirectory.standardizedFileURL
            .appendingPathComponent(safeFolderName, isDirectory: true)
        try ensureDestinationAvailable(destination)

        for _ in 0..<entryCount {
            let remaining = payload.count - offset
            let (relativePath, fileData) = try readBlob(payload, offset: &offset, remaining: remaining)
            let safePath = try sanitizeRelativePath(relativePath)
            let fileURL = destination.appendingPathComponent(safePath)
            guard isContainedInDirectory(fileURL, directory: destination) else {
                throw EnclaveError.invalidFormat
            }

            let parent = fileURL.deletingLastPathComponent()
            if !parent.path.isEmpty {
                try FileManager.default.createDirectory(at: parent, withIntermediateDirectories: true)
            }
            try fileData.write(to: fileURL, options: .atomic)
        }

        guard offset == payload.count else {
            throw EnclaveError.invalidFormat
        }
    }

    private static func ensureDestinationAvailable(_ destination: URL) throws {
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: destination.path, isDirectory: &isDirectory) else {
            try FileManager.default.createDirectory(at: destination, withIntermediateDirectories: true)
            return
        }

        guard isDirectory.boolValue else {
            throw EnclaveError.destinationExists
        }

        let contents = try FileManager.default.contentsOfDirectory(atPath: destination.path)
        guard contents.isEmpty else {
            throw EnclaveError.destinationExists
        }
    }

    private static func sanitizeFolderName(_ name: String) throws -> String {
        guard !name.isEmpty, !name.contains("/"), !name.contains("\\") else {
            throw EnclaveError.invalidFormat
        }
        let base = (name as NSString).lastPathComponent
        guard !base.isEmpty, base != ".", base != ".." else {
            throw EnclaveError.invalidFormat
        }
        guard !base.contains("..") else {
            throw EnclaveError.invalidFormat
        }
        return base
    }

    private static func relativePath(from root: URL, to file: URL) throws -> String {
        let rootPath = root.standardizedFileURL.path
        let filePath = file.standardizedFileURL.path
        let prefix = rootPath + "/"
        guard filePath.hasPrefix(prefix) else {
            throw EnclaveError.invalidFormat
        }
        return try sanitizeRelativePath(String(filePath.dropFirst(prefix.count)))
    }

    private static func sanitizeRelativePath(_ path: String) throws -> String {
        guard !path.isEmpty, !path.hasPrefix("/"), !path.contains("\\") else {
            throw EnclaveError.invalidFormat
        }
        guard path.utf8.count <= maxPathLength else {
            throw EnclaveError.invalidFormat
        }

        let parts = path.split(separator: "/", omittingEmptySubsequences: false).map(String.init)
        guard !parts.isEmpty,
              !parts.contains(".."),
              !parts.contains("."),
              !parts.contains(where: { $0.isEmpty }) else {
            throw EnclaveError.invalidFormat
        }
        return parts.joined(separator: "/")
    }

    private static func isContainedInDirectory(_ fileURL: URL, directory: URL) -> Bool {
        let fileComponents = fileURL.standardizedFileURL.pathComponents
        let directoryComponents = directory.standardizedFileURL.pathComponents
        guard fileComponents.count >= directoryComponents.count else {
            return false
        }
        return Array(fileComponents.prefix(directoryComponents.count)) == directoryComponents
    }

    private static func appendUInt32(_ value: UInt32, to data: inout Data) {
        var be = value.bigEndian
        withUnsafeBytes(of: &be) { data.append(contentsOf: $0) }
    }

    private static func readUInt32(_ data: Data, offset: inout Int) throws -> UInt32 {
        guard offset + 4 <= data.count else {
            throw EnclaveError.invalidFormat
        }
        let bytes = [UInt8](data[offset..<(offset + 4)])
        offset += 4
        return (UInt32(bytes[0]) << 24) | (UInt32(bytes[1]) << 16)
            | (UInt32(bytes[2]) << 8) | UInt32(bytes[3])
    }

    private static func appendBlob(_ path: String, data fileData: Data, to archive: inout Data) {
        let pathData = Data(path.utf8)
        precondition(pathData.count <= UInt32.max)
        appendUInt32(UInt32(pathData.count), to: &archive)
        archive.append(pathData)
        var size = UInt64(fileData.count).bigEndian
        withUnsafeBytes(of: &size) { archive.append(contentsOf: $0) }
        archive.append(fileData)
    }

    private static func readBlob(_ data: Data, offset: inout Int, remaining: Int) throws -> (String, Data) {
        guard remaining >= 4 else {
            throw EnclaveError.invalidFormat
        }

        let pathLength = Int(try readUInt32(data, offset: &offset))
        guard pathLength > 0,
              pathLength <= maxPathLength,
              offset + pathLength + 8 <= data.count else {
            throw EnclaveError.invalidFormat
        }

        let pathData = data[offset..<(offset + pathLength)]
        offset += pathLength
        guard let path = String(data: pathData, encoding: .utf8) else {
            throw EnclaveError.invalidFormat
        }

        let sizeBytes = [UInt8](data[offset..<(offset + 8)])
        offset += 8
        var fileLength64: UInt64 = 0
        for b in sizeBytes { fileLength64 = (fileLength64 << 8) | UInt64(b) }
        // Zero-length entries are valid — empty files must round-trip faithfully.
        // Validate as UInt64 before narrowing to Int so a corrupt size field
        // throws cleanly instead of trapping the Int conversion.
        guard fileLength64 <= UInt64(maxEntrySize) else {
            throw EnclaveError.invalidFormat
        }
        let fileLength = Int(fileLength64)
        guard offset + fileLength <= data.count else {
            throw EnclaveError.invalidFormat
        }

        let fileData = Data(data[offset..<(offset + fileLength)])
        offset += fileLength
        return (path, fileData)
    }
}